src/django_upgrade/fixers/admin_register.py

```diff
@@ -0,0 +1,109 @@
+
+fixer = Fixer(
+    __name__,
+    min_version=(1, 7),
+)
+
+# Keep track of classes that could be decorated with `@admin.register()`
+# For each class name, store the associated custom ModelAdmin class
+# inferred from eligible `admin.site.register` calls.
+class_to_decorate: MutableMapping[State, dict[str, set[str]]] = WeakKeyDictionary()
+
+
+@fixer.register(ast.ClassDef)
+def visit_ClassDef(
+    state: State,
+    node: ast.ClassDef,
+    parent: ast.AST,
+) -> Iterable[tuple[Offset, TokenFunc]]:
+    if "admin" in state.from_imports["django.contrib"] and not node.decorator_list:
+        class_to_decorate.setdefault(state, {})[node.name] = set()
+        yield ast_start_offset(node), partial(
+            update_class_def,
+            node=node,
+            state=state,
+        )
+
+
+def update_class_def(
+    tokens: list[Token], i: int, *, node: ast.ClassDef, state: State
+) -> None:
+    arg_kwargs = class_to_decorate.get(state, {}).pop(node.name, set())
+    if len(arg_kwargs) == 1:
+        j, indent = extract_indent(tokens, i)
+        insert(
+            tokens,
+            j,
+            new_src=f"{indent}@admin.register({arg_kwargs.pop()})\n",
+        )
+
+
+@fixer.register(ast.Call)
+def visit_Call(
+    state: State,
+    node: ast.Call,
+    parent: ast.AST,
+) -> Iterable[tuple[Offset, TokenFunc]]:
+    if (
+        "admin" in state.from_imports["django.contrib"]
+        and isinstance(node.func, ast.Attribute)
+        and node.func.attr == "register"
+        and isinstance(node.func.value, ast.Attribute)
+        and node.func.value.attr == "site"
+        and isinstance(node.func.value.value, ast.Name)
+        and node.func.value.value.id == "admin"
+        and not node.keywords
+        and (
+            len(node.args) == 2
+            and isinstance(node.args[0], ast.Name)
+            and isinstance(node.args[1], ast.Name)
+        )
+    ):
+        admin_model_name = node.args[1].id
+        if admin_model_name in class_to_decorate.get(state, {}).keys():
+            class_to_decorate[state][admin_model_name].add(node.args[0].id)
+            yield ast_start_offset(node), partial(
+                erase_register_node,
+                node=parent,
+                admin_model_name=admin_model_name,
+                state=state,
+            )
+
+    elif (
+        "admin" in state.from_imports["django.contrib"]
+        and isinstance(node.func, ast.Name)
+        and node.func.id == "super"
+        and isinstance(parent, ast.Attribute)
+        and parent.attr in {"__init__", "__new__"}
+        and len(node.args) == 2
+        and isinstance(node.args[0], ast.Name)
+        and isinstance(node.args[1], ast.Name)
+    ):
+        # Cannot convert class using py2 style `super(MyAdmin, self)`
+        # in its `__init__` or `__new__` method.
+        # https://docs.djangoproject.com/en/dev/ref/contrib/admin/#the-register-decorator
+        class_to_decorate[state].popitem()
```

Comment:
this isn't exactly precise, it's not guaranteed the `__init__` / `__new__` is part of the popped admin class, and special cases like multiple `super()` calls will cause multiple admin classes to be popped, e.g.

```
class HamAdmin(...):
    ...


class SpamAdmin(...):
    def __init__(self, *args, **kwargs):
        if sys.version_info >= (3, 10):
			# something
            super(SpamAdmin, self).__init__(*args, **kwargs)
		else:
			# something else
            super(SpamAdmin, self).__init__(*args, **kwargs)
```

I'd feel more confident if the code checked for the special case in `visit_ClassDef`. It could walk the class def body, looking for `__init__` / `__new__` function defs, and then check their bodies for any unsupported `super()` calls. If there's a match, no need to ever add the class to the to-decorate list.

Reply:
Oh you're right! Didn't thought about these cases. Will look into that tomorrow, should not be too difficult

Reply:
I've moved the implementation inside the `visit_ClassDef`, this is not perfect because it's walking the tree another time but only when specific conditions are met so it's probably fine.

Reply:
Yeah I think the overhead should be fine.